kernel: fix error propagation for device deferred initialization

ljd42 opened the pull request:
This fix propagates the failure code returned from the device initialization function to the application when deferred initialization is used. Previously, `device_init()` mistakenly returned `errno` instead of `-errno`.

A test has been added to test.kernel.device to verify the changes  (fail before the fix, pass after).

Fix: #91095

andyross commented on kernel/init.c:
```diff
@@ -311,24 +311,19 @@ static int do_device_init(const struct device *dev)
 	int rc = 0;
 
+	/* device initialization has been invoked */
+	dev->state->initialized = true;
```
Seems questionable to me to flag this **before** making the init call.  Are you sure this is what the docs require?  Maybe the docs are just wrong?

peter-mitsis replied:
It is not yet obvious to me why the setting of this `initialized` field to true was moved here. Doing so, does raise a potential race condition concern here to me.

The routine `device_is_ready()` checks both `initialized` and `init_res`. If there is a context switch between setting `initialized` to true and a failing call to dev->ops.init(), device_is_ready() might mistakenly think that the device is ready, when it has not yet finished initialization.

Is this possible? I'm not sure, but I wanted to raise this just in case.

andyross reviewed (commented):
Question on the initialized flag semantics.  Also pedantry: you want to introduce the test after the fix in the commit sequence, otherwise the PR introduces a bisection failure.

peter-mitsis commented on kernel/init.c:
```diff
@@ -311,24 +311,19 @@ static int do_device_init(const struct device *dev)
-				rc = UINT8_MAX;
-			}
-			dev->state->init_res = rc;
+			int res = (rc < 0) ? -rc : rc;
+
+			dev->state->init_res = (res <= UINT8_MAX) ? res : UINT8_MAX;
```
Two things jump out at me here.

The first is that unless I am missing something, this just gives the same behavior as the original code. If that is true, how does that fix anything?

The second is a logical corner case. If the return code `rc` from `dev->ops.init()` is ever INT_MIN (0x80000000), then `-rc` is the same as `rc`, and the value that we save will be 0x0--a false success path. Would it be worthwhile to have an assert that covers that?

peter-mitsis replied:
Aaah ... I finally see what my eyes kept glazing over on the previous pass with regards to the signedness of `rc`.

peter-mitsis reviewed (commented):
A couple of comments--hopefully they are useful.

cfriedt reviewed (commented):
Sorry - I started a review of this quite a while ago but didn't finish. Please see my comments as well.

ljd42 commented:
> Question on the initialized flag semantics. Also pedantry: you want to introduce the test after the fix in the commit sequence, otherwise the PR introduces a bisection failure.

For bugs, introducing the test first, then the fix, has the advantage to give an audit trail. (fail before the fix, pass after). This can be useful for certification process/audit (IEC 61508 and the like).

If that's a problem from a git/zephyr way of working, happy to reorder... I have anyway some reviews findings from Cris to include for the test :wink:

ljd42 commented on kernel/init.c:
```diff
@@ -311,24 +311,19 @@ static int do_device_init(const struct device *dev)
 {
 	int rc = 0;
 
+	/* device initialization has been invoked */
+	dev->state->initialized = true;
+
```
yes, yes and yes.

ljd42 commented on tests/kernel/device/src/main.c:
```diff
@@ -415,6 +422,38 @@ ZTEST(device, test_deferred_init)
+
+	zassert_false(device_is_ready(dev));
+	ret = device_init(dev);
+	zassert_true(ret == -EIO);
```
Damn, I should have come to it myself.... But I guess, there's a few more tests there we could improve too :wink:

ljd42 commented on tests/kernel/device/src/main.c:
```diff
@@ -415,6 +422,38 @@ ZTEST(device, test_deferred_init)
+	ret = device_init(dev);
+	zassert_true(ret == -EIO);
+	zassert_false(device_is_ready(dev));
+	zassert_true(dev->state->init_res == EIO);
```
ACK

ljd42 commented on tests/kernel/device/src/main.c:
```diff
@@ -415,6 +422,38 @@ ZTEST(device, test_deferred_init)
+	zassert_true(dev->state->init_res == EIO);
+
+	ret = device_init(dev);
+	zassert_true(ret == -EALREADY);
+	zassert_true(dev->state->init_res == EIO);
```
ACK

ljd42 committed:
kernel: fix error propagation for device deferred initialization

This fix makes sure that do_device_init() returns a negative value if
the device's initialization failed. Previously, it mistakely returned
+errno instead of -errno.

This oversight happened during the refactoring of z_sys_init_run_level()
to support deferred initialization, from which most of do_device_init()
code derives. The rc value computed and stored in dev->state->init_res
is the POSITIVE value of the resulting errno. Returning rc therefore
breaks the convention of a negative value to signal failure.

Signed-off-by: Loic Domaigne <tech@domaigne.com>

ljd42 committed:
tests: kernel: device: Add tests for deferred init failure

These tests verify the expected returned value and device state for
error cases (-EALREADY, -errno) conform to the device_init()
documentation.

Signed-off-by: Loic Domaigne <tech@domaigne.com>

ljd42 commented:
> Right, but it breaks "git bisect" by introducing an unrelated/phantom failure in the tree, which is an actual software tool we need to run with some regularity. Human audits can always make do with an extra patch reversion or whatever. Please reorder :)

Yes. Reordering as already happened.

ljd42 commented on kernel/init.c:
```diff
@@ -313,20 +313,16 @@ static int do_device_init(const struct device *dev)
-			}
-			dev->state->init_res = rc;
+			__ASSERT(rc != INT_MIN, "init(dev) failed, but init_res will be 0");
+			int res = (rc < 0) ? -rc : rc;
+
+			dev->state->init_res = (res <= UINT8_MAX) ? res : UINT8_MAX;
```
@cfriedt : Ok, this makes the reading easier. But the code is located in the kernel/  folder, and none of the files there use stdlib.h. So, I'm not sure if it's a good idea. 

A `CLAMP()` macro is available  in sys/util.h, and is used in sched.c for instance. There's no `abs()` however.

ljd42 replied:
I am going to implement your proposed changes, Chris, as it improves the code reading.

ljd42 commented:
This PR will be now for Zephyr v4.3.

cfriedt reviewed (dismissed):
LGTM - the only thing I didn't notice earlier is that `<zephyr/sys/util.h>` isn't explicitly included, but if it builds, it builds.

andyross commented on kernel/init.c:
```diff
@@ -41,6 +41,11 @@
 
+/* avoid pulling in stdlib */
+#undef _abs
+#define _abs(x) ((x) < 0 ? (-(x)) : (x))
```
Style: not loving this.  That `#undef` seems like an admission that there is a known collision with another symbol, in which case this is a bug.  If it's not colliding, letting the preprocessor detect and warn on the redefinition is the correct way to be sure continues not to collide. 

Also: this isn't actually used to compute an "absolute value" in a semantic sense.  It's to handle the variant signedness conventions of an error return value, which are logically enumerates and not integers on which you do math.  IMHO the logic belongs below where you're processing the error value, ideally with a comment explaining e.g. "init_res stores the POSITIVE value of the resulting errno".

ljd42 replied:
> That #undef seems like an admission that there is a known collision with another symbol [...]

I have checked that there's no collision before making the change...  While the intend is clear (local override), I wondered if that #undef was a common technique to use within the Zephyr kernel :sweat_smile:  It is used elsewhere, and I trusted Chris suggestion.

I'm absolutely with you on this. A redefinition causes indeed a build error. That's the way to go!

> Also: this isn't actually used to compute an "absolute value" in a semantic sense.  [...] 

Yes, I agree.

andyross reviewed (changes requested):
Sorry, some nits on the error code handling that rise just to my -1 threshold.  The preprocessor one is the only fatal one though, I'm willing to be convinced on the translation code for the number.

ljd42 commented on kernel/init.c:
```diff
@@ -313,20 +318,14 @@ static int do_device_init(const struct device *dev)
-			}
-			dev->state->init_res = rc;
+			__ASSERT(rc != INT_MIN, "init(dev) failed, but init_res will be 0");
+			dev->state->init_res = CLAMP(_abs(rc), 0, UINT8_MAX);
```
At first glance, the  suggested change seemed to require less mental cycles to process than the first version with the two conditional expressions...

>  An overflow is an error no matter what.  [...] Truncating is no less an error.

I agree with your line of reasoning. But note that the code currently in tree also truncates to UINT8_MAX:
https://github.com/zephyrproject-rtos/zephyr/blob/fdd02c5105d2a40cf6410cf89ccd22c43a48399e/kernel/init.c#L322-L328

From what I can see, it does so because dev->state->init_res is an uint8_t. It's better to store 255 in case of an overflow - even if that is bogus - than 0 if the value happens to be a multiple of 256.

Archeological note: this piece of code is present since 2021, and has been introduced with Zephyr 2.6. See device.c::z_sys_init_run_level()

ljd42 commented:
I found a radical solution that minimizes the code change, and should address all review comments discussed so far. I hope that works for you.

bjarki-andreasen reviewed (approved):
LGTM

dkalowsk reviewed (approved):
approving the changes.  Does this need a call out in the release notes though?  Seems like a rather big change.

ljd42 commented:
@dkalowsk : yes, absolutely. This needs to be called out in the **Release note and migration guide**. This has been analyzed in the linked issue, before we decide how to go about it. Just point me to the doc to edit, and I'll take care of it. 

Quoting from #91095:

> Of course, the change will potentially impact code **using deferred initialization**.  I see 3 groups:
>
>  -  group 1: does not care to check the return value of device_init() => no impact. But they should really start checking 😉
>   - group 2: implement the error checking following the spec. Error condition is never detected => positive impact as error > condition are now detected that weren't previously.
>  - group 3: has implemented a workaround for the faulty error condition => they need to remove their workaround. We should put a note in the migration guide.

cfriedt commented:
@andyross @peter-mitsis - I think we need an approval from either of you as assignee

dkalowsk commented:
> @dkalowsk : yes, absolutely. This needs to be called out in the **Release note and migration guide**. This has been analyzed in the linked issue, before we decide how to go about it. Just point me to the doc to edit, and I'll take care of it.

@ljd42  check under `docs/releases/release-notes-4.3.txt` and maybe also the migration.txt for the release.

andyross reviewed (approved):
Had to be summoned from discord.  No more complaints.
